Spotfleet Launch Template and On-Demand Capacity Support

lossanarch opened the pull request:
Updates the spotfleet_request resource to support the new launch_template resource and on-demand capacity.

Changes proposed in this pull request:

* launch_template_configs

The new block looks like this, following the structure from the go sdk as closely as possible:
```hcl
launch_template_configs {
    launch_template_specification {
      id = "${aws_launch_template.ecs_instance.id}"
      version = "${aws_launch_template.ecs_instance.latest_version}"
    }
    overrides {
      instance_type = "t2.micro"
      spot_price = "0.05"
    }
  }
```

* on_demand_target_capacity

Allows specifying a minimum portion of capcity to be filled by on-demand instances.

Any other tips & suggestions are appreciated.

Output from acceptance testing:

```
=== RUN   TestAccAWSSpotFleetRequest_associatePublicIpAddress
--- PASS: TestAccAWSSpotFleetRequest_associatePublicIpAddress (298.31s)
=== RUN   TestAccAWSSpotFleetRequest_launchTemplate
--- PASS: TestAccAWSSpotFleetRequest_launchTemplate (385.74s)
=== RUN   TestAccAWSSpotFleetRequest_launchTemplateWithOnDemandCapacity
--- PASS: TestAccAWSSpotFleetRequest_launchTemplateWithOnDemandCapacity (118.89s)
=== RUN   TestAccAWSSpotFleetRequest_launchTemplateWithOverrides
--- PASS: TestAccAWSSpotFleetRequest_launchTemplateWithOverrides (258.26s)
=== RUN   TestAccAWSSpotFleetRequest_launchTemplateToLaunchSpec
--- PASS: TestAccAWSSpotFleetRequest_launchTemplateToLaunchSpec (637.84s)
=== RUN   TestAccAWSSpotFleetRequest_launchSpecToLaunchTemplate
--- PASS: TestAccAWSSpotFleetRequest_launchSpecToLaunchTemplate (521.70s)
=== RUN   TestAccAWSSpotFleetRequest_instanceInterruptionBehavior
--- PASS: TestAccAWSSpotFleetRequest_instanceInterruptionBehavior (325.34s)
=== RUN   TestAccAWSSpotFleetRequest_changePriceForcesNewRequest
--- PASS: TestAccAWSSpotFleetRequest_changePriceForcesNewRequest (635.60s)
=== RUN   TestAccAWSSpotFleetRequest_lowestPriceAzOrSubnetInRegion
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceAzOrSubnetInRegion (325.68s)
=== RUN   TestAccAWSSpotFleetRequest_lowestPriceAzInGivenList
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceAzInGivenList (352.56s)
=== RUN   TestAccAWSSpotFleetRequest_lowestPriceSubnetInGivenList
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceSubnetInGivenList (262.74s)
=== RUN   TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameAz
--- PASS: TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameAz (318.36s)
=== RUN   TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameSubnet
--- PASS: TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameSubnet (333.31s)
=== RUN   TestAccAWSSpotFleetRequest_overriddingSpotPrice
--- PASS: TestAccAWSSpotFleetRequest_overriddingSpotPrice (391.70s)
=== RUN   TestAccAWSSpotFleetRequest_withoutSpotPrice
--- PASS: TestAccAWSSpotFleetRequest_withoutSpotPrice (385.29s)
=== RUN   TestAccAWSSpotFleetRequest_withWeightedCapacity
--- PASS: TestAccAWSSpotFleetRequest_withWeightedCapacity (326.93s)
=== RUN   TestAccAWSSpotFleetRequest_withEBSDisk
--- PASS: TestAccAWSSpotFleetRequest_withEBSDisk (311.59s)
=== RUN   TestAccAWSSpotFleetRequest_withTags
--- PASS: TestAccAWSSpotFleetRequest_withTags (325.39s)
=== RUN   TestAccAWSSpotFleetRequest_placementTenancy
--- PASS: TestAccAWSSpotFleetRequest_placementTenancy (107.13s)
=== RUN   TestAccAWSSpotFleetRequest_diversifiedAllocation
--- PASS: TestAccAWSSpotFleetRequest_diversifiedAllocation (363.19s)
=== RUN   TestAccAWSSpotFleetRequest_WithELBs
--- PASS: TestAccAWSSpotFleetRequest_WithELBs (350.88s)
=== RUN   TestAccAWSSpotFleetRequest_WithTargetGroups
--- PASS: TestAccAWSSpotFleetRequest_WithTargetGroups (443.07s)
```

The above was the full set of spot fleet resource tests.
Of the above tests, these are the new ones:

```
=== RUN   TestAccAWSSpotFleetRequest_launchTemplate
--- PASS: TestAccAWSSpotFleetRequest_launchTemplate (385.74s)
=== RUN   TestAccAWSSpotFleetRequest_launchTemplateWithOnDemandCapacity
--- PASS: TestAccAWSSpotFleetRequest_launchTemplateWithOnDemandCapacity (118.89s)
=== RUN   TestAccAWSSpotFleetRequest_launchTemplateWithOverrides
--- PASS: TestAccAWSSpotFleetRequest_launchTemplateWithOverrides (258.26s)
=== RUN   TestAccAWSSpotFleetRequest_launchTemplateToLaunchSpec
--- PASS: TestAccAWSSpotFleetRequest_launchTemplateToLaunchSpec (637.84s)
=== RUN   TestAccAWSSpotFleetRequest_launchSpecToLaunchTemplate
--- PASS: TestAccAWSSpotFleetRequest_launchSpecToLaunchTemplate (521.70s)
```

kkopachev commented:
> The numbers in the resource attr paths seem to be randomly generated and change every time I run the tests, so I can never get them to match my output assertions.

Looks like they calculated here and should not change if input config did not changed.
https://github.com/terraform-providers/terraform-provider-aws/blob/cca70b5b510958ea5f34412760143befa03573e0/resource_aws_spot_fleet_request.go#L996-L1009

lossanarch commented:
Ok, barring feedback and change requests this is now done with all spot fleet tests passing.

@kkopachev I added a hashing function for the launch template set but tbh I'm still a little fuzzy on whether it was necessary.
Your comment about the hash staying the same if the config didn't change was very helpful - I realised I was using the launch template id which would change each generation, so I switched to using name which made the hashes predictable.
Thanks again!

kkopachev commented:
@lossanarch Could you rebase to resolve conflicts?

lossanarch commented:
Sorry @kkopachev, for some reason I thought it had to be done by someone with write access. Please excuse my inexperience.
Looks clear now.

fredsig commented:
Is this going to be included in the upcoming 1.29.0? Thanks.

lossanarch commented:
Is there anything else I need to address with this one?

lossanarch commented:
Any word on when this might be merged?

Zeerg commented:
Can we get this in the next release please?

Ninir commented:
@lossanarch Hey, could you rebase this branch? will have a look at it as soon as it's resolved :)
Thanks!

siwyd commented:
@lossanarch Definitely a big fan of this getting merged!

lossanarch commented:
Rebase complete.

Getting this merged would make my life a whole lot easier as we use this in production, I'd appreciate if you could take a look @Ninir.

Ninir commented:
I will this week :) 👍

lossanarch commented on aws/resource_aws_spot_fleet_request_test.go:
```diff
@@ -937,24 +1248,35 @@ resource "aws_iam_role" "test-role" {
+
+    launch_template_configs {
+      launch_template_specification {
+        name = "${aws_launch_template.foo.name}"
```
I did try with ids but because they are generated each time I found it impossible to predict the resource hash required to check the results; any ideas?

lossanarch commented:
Thanks @ninir I will get into all those shortly!

Just re splitting into two, the on demand capacity feature is dependent upon the launch template support - do you still want me to split? Maybe split and note on the on demand PR that it’s dependent on the launch template one?

Ninir commented:
@lossanarch If we could have one merged and then the other one, that would be preferable yes :)

lossanarch commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -360,6 +454,14 @@ func resourceAwsSpotFleetRequest() *schema.Resource {
 			},
 		},
+		CustomizeDiff: customdiff.Sequence(
+			customdiff.ComputedIf("launch_template_configs.0.launch_template_specification.0.id", func(diff *schema.ResourceDiff, meta interface{}) bool {
```
I actually stole this from the launch template implementation in the autoscaling group - I've commented it as how I understand it works, hopefully I'm right. 😬

lossanarch commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -619,6 +795,32 @@ func resourceAwsSpotFleetRequestCreate(d *schema.ResourceData, meta interface{})
+		spotFleetConfig.LaunchTemplateConfigs, err = buildLaunchTemplateConfigs(d, meta)
+		if err != nil {
+			return err
+		}
```
Cleaned and logic standardised. Commit 127425f

lossanarch commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -703,10 +905,22 @@ func resourceAwsSpotFleetRequestCreate(d *schema.ResourceData, meta interface{})
-						fmt.Errorf("Error creating Spot fleet request, retrying: %s", err))
+					switch {
+					case strings.Contains(awsErr.Message(), "LaunchTemplateSpecification"):
+						return resource.NonRetryableError(err)
```
Comment added, commit 95e2fc6

lossanarch commented on website/docs/r/spot_fleet_request.html.markdown:
```diff
@@ -90,23 +164,25 @@ Most of these arguments directly correspond to the
 * `target_capacity` - The number of units to request. You can choose to set the
   target capacity in terms of instances or a performance characteristic that is
   important to your application workload, such as vCPUs, memory, or I/O.
+* `on_demand_target_capacity` - (Optional) Allows specification of a minimum value of the `target_capacity` to be filled by full price on-demand instance to ensure a minimum capacity of the fleet no matter spot instance availability. Not supported by `launch_specification`, must be used with `launch_template_configs`.
```
Hm I agree this is not my best copy, but I'm not a huge fan of the aws doc for it either. It's a bit too generic and I'm not sure "add[ing] capacity later" is even accurate - as I understand the api it requires a new request to change the on-demand capacity (as opposed to the regular target_capacity, which can be modified). How about a hybrid:

> The number of full-price On-Demand units to request in addition to the spot instance capacity. Use this to guarantee a minimum capacity of the fleet even in the event that the spot requests cannot be filled. Capacity is defined in the same units as configured in `target_capacity` such as vCPUs, memory, or I/O. Not supported by `launch_specification`, must be used with `launch_template_configs`.

How about something like that?

I'll move this out with the rest of the on-demand target capacity feature soon, when I get around to splitting them.

lossanarch commented on aws/resource_aws_spot_fleet_request_test.go:
```diff
@@ -144,13 +158,328 @@ func TestAccAWSSpotFleetRequest_iamInstanceProfileArn(t *testing.T) {
+					resource.TestCheckResourceAttr(
+						"aws_spot_fleet_request.foo", "launch_specification.#", "1"),
+					// resource.TestCheckResourceAttr(
+					// 	"aws_spot_fleet_request.foo", "launch_template_configs.#", "0"),
```
Sorry, should really be neater. Fixed in commit f494fca

lossanarch commented on aws/resource_aws_spot_fleet_request_test.go:
```diff
@@ -144,13 +158,328 @@ func TestAccAWSSpotFleetRequest_iamInstanceProfileArn(t *testing.T) {
+					resource.TestCheckResourceAttr(
+						"aws_spot_fleet_request.foo", "launch_specification.#", "1"),
+					// resource.TestCheckResourceAttr(
+					// 	"aws_spot_fleet_request.foo", "launch_template_configs.#", "0"),
```
Removed. Commit 247d99e

lossanarch commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -1198,6 +1453,24 @@ func deleteSpotFleetRequest(spotFleetRequestID string, terminateInstances bool,
+			if len(iresp.Reservations) == 0 {
+				log.Printf("[DEBUG] Active instance count is %d for Spot Fleet Request (%s), but instances have terminated, removing", len(resp.ActiveInstances), spotFleetRequestID)
+				return nil
+			} else {
```
Good point, fixed in 2e33ef4.

lossanarch commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -273,12 +276,103 @@ func resourceAwsSpotFleetRequest() *schema.Resource {
 				Required: true,
 				ForceNew: false,
 			},
+			"on_demand_target_capacity": {
```
I've now split this out

lossanarch commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -598,19 +700,93 @@ func buildAwsSpotFleetLaunchSpecifications(
+	_, launchTemplateConfigsOk := d.GetOk("launch_template_configs")
+
+	if !launchSpecificationOk && !launchTemplateConfigsOk {
+		return fmt.Errorf("One of `launch_specification` or `launch_template` must be set for an a fleet request")
```
Resolved

lossanarch commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -703,10 +905,22 @@ func resourceAwsSpotFleetRequestCreate(d *schema.ResourceData, meta interface{})
+
+					// Assuming this doesn't work; we can get the same error code for reasons other than a lack of IAM consistency:
+					//return resource.RetryableError(
+					//	fmt.Errorf("[WARN] Error creating Spot fleet request, retrying: %s", err))
```
Removed

lossanarch commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -934,9 +1148,49 @@ func resourceAwsSpotFleetRequestRead(d *schema.ResourceData, meta interface{}) e
+	return overrideSet
+}
+
+func overrideToMap(override *ec2.LaunchTemplateOverrides) map[string]interface{} {
```
Yep, probably right. Fixed in commit d7fb944.

lossanarch commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -273,12 +276,103 @@ func resourceAwsSpotFleetRequest() *schema.Resource {
+				Type:          schema.TypeSet,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"launch_specification"},
```
Added, commit aa79ad2

lossanarch commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -273,12 +276,103 @@ func resourceAwsSpotFleetRequest() *schema.Resource {
+										Type:     schema.TypeString,
+										Optional: true,
+										ForceNew: true,
+										//ValidateFunc:  validateLaunchTemplateName,
```
Removed

lossanarch commented on website/docs/r/spot_fleet_request.html.markdown:
```diff
@@ -131,6 +207,33 @@ across different markets and instance types.
+
+* `id` - The ID of the launch template. Conflicts with `name`.
+* `name` - The name of the launch template. Conflicts with `id`.
+* `version` - (Optional) Template version. Unlike the autoscaling equivalent, does not support `$Latest` or `$Default`, so use the launch_template resource's attribute, e.g. `"${aws_launch_template.foo.latest_version}"`. Wil use default version if omitted.
```
Fixed

lossanarch commented:
Ok @Ninir I believe that's everything. On-demand capacity has been removed to be included in a separate PR.

Chuwiey commented:
Would love to see this merged, along with a modification for the requirement of specifying the instance type. According to AWS's docs - it isn't a required field?

Ninir commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -273,12 +276,103 @@ func resourceAwsSpotFleetRequest() *schema.Resource {
+										Optional:      true,
+										Computed:      true,
+										ForceNew:      true,
+										ConflictsWith: []string{"launch_template_configs.0.launch_template_specification.0.name"},
```
@lossanarch Not sure to see the test on this one, were you able to check it?

lossanarch replied:
Hey I believe this one is covered by `TestAccAWSSpotFleetRequest_launchTemplateConflictLaunchSpecification`, line 166 of the tests file.

lossanarch commented on aws/resource_aws_spot_fleet_request_test.go:
```diff
@@ -83,12 +83,452 @@ func TestAccAWSSpotFleetRequest_associatePublicIpAddress(t *testing.T) {
+	})
+}
+
+func TestAccAWSSpotFleetRequest_launchTemplateConflictLaunchSpecification(t *testing.T) {
```
Hi @Ninir , the test I believe you're referring to is here

lossanarch commented:
Huge apologies to anyone waiting on this, unfortunately I just have very little time to devote to it anymore.

@Ninir I've addressed your question on the test for the launch_template/launch_spec conflict, let me know if there's anything else.

@Chuwiey the required instance_type you must be referring to must be this one:
```
"instance_type": {
    Type:     schema.TypeString,
    Required: true,
    ForceNew: true,
},
```
which is the one defined for the **launch specification**.

My one is defined in the **launch template** stanza, and is optional:
```
"instance_type": {
    Type:     schema.TypeString,
    Optional: true,
    ForceNew: true,
},
```

I think changing the launch_specification instance_type `Required` flag is probably out of scope for this PR as I purposefully tried to avoid any changes to the launch specification to keep this PR as small as possible (as it is, it's still huge).

lossanarch commented:
Ok I've also rebased against master again. If you can approve at this point please please merge asap, for the love of all that's holy please don't make me rebase it again 😭

oliversalzburg commented:
How about getting this finally merged? What is the point of having a launch template if you can't use it?

ddonaghy-r7 commented:
Any status on having this PR merged?

oliversalzburg commented:
Just FYI, we just went with an `aws_spot_fleet_request` instead. For our purposes, it allowed us to construct the same result we previously achieved with launch templates.

thaiphv commented:
Any update on this?

thaiphv commented:
@Ninir do you have any plan to review and merge this PR?

alexejk commented:
Really looking forward to this one being merged soon.

thaiphv commented:
Why the silence? Should the PR be assigned to another reviewer?

Ninir commented:
Sorry folks, I've been relocating from one country another and I need to get back in shape to contributing here.
I can ask @bflad to have a look at it, but I think he's busy on other things right now. I'll do my best in the coming days to make that happen, and again sorry for the latency here.

Ninir commented on aws/resource_aws_spot_fleet_request_test.go:
```diff
@@ -83,12 +83,452 @@ func TestAccAWSSpotFleetRequest_associatePublicIpAddress(t *testing.T) {
+					resource.TestCheckResourceAttr(
+						"aws_spot_fleet_request.foo", "launch_template_configs.#", "1"),
+					resource.TestCheckResourceAttr(
+						"aws_spot_fleet_request.foo", "launch_template_configs.4018047529.launch_template_specification.#", "1"),
```
Not for me, I'm gonna check this one once running acceptance tests with TF_SCHEMA_PANIC_ON_ERROR=1 to be sure it's not a false positive

brianpham commented:
@lossanarch Any updates on this? Was trying to follow along where the progress of this issue is, but not sure since it haven't been updated in awhile.

gammasts commented:
@lossanarch are you able to continue this PR? I think it's super valuable since it brings parity to spot fleet requests especially because AWS seems to be going down the path of supporting launch templates for all the other resources (EC2 Fleet, Auto Scaling).

@bflad I noticed you had updated the milestones previously but somehow it got removed and GitHub made no mention of it. Anything you can do to assist shepherding this PR? I'm getting worried that this one has lost all of it's steam.

lossanarch commented on aws/resource_aws_spot_fleet_request.go:
```diff
@@ -273,7 +276,88 @@ func resourceAwsSpotFleetRequest() *schema.Resource {
+										Optional:     true,
+										Computed:     true,
+										ForceNew:     true,
+										ValidateFunc: validation.StringLenBetween(1, 255),
```
I think this may have meant to be for a "version_description" attr, which is limited to 255 characters. The api lists "version" as type long though, so this definitely needs fixing.

lossanarch commented on website/docs/r/spot_fleet_request.html.markdown:
```diff
@@ -131,6 +206,33 @@ across different markets and instance types.
+
+* `id` - The ID of the launch template. Conflicts with `name`.
+* `name` - The name of the launch template. Conflicts with `id`.
+* `version` - (Optional) Template version. Unlike the autoscaling equivalent, does not support `$Latest` or `$Default`, so use the launch_template resource's attribute, e.g. `"${aws_launch_template.foo.latest_version}"`. Will use default version if omitted.
```
I always prefer the passive voice in documentation personally but if you'd rather it this way that's fine

lossanarch commented on website/docs/r/spot_fleet_request.html.markdown:
```diff
@@ -131,6 +206,33 @@ across different markets and instance types.
+
+### Overrides
+
+* `instance_type` - (Optional) The type of instance to request.
```
This sounds like a good idea but so far I've not been able to use 0.12 as not all providers I'm using are compatible. I'll need to get acquainted with it before I can do this.

lossanarch commented on website/docs/r/spot_fleet_request.html.markdown:
```diff
@@ -131,6 +206,33 @@ across different markets and instance types.
+* `weighted_capacity` - (Optional) The capacity added to the fleet by a fulfilled request.
+* `subnet_id` - (Optional) The subnet in which to launch the requested instance.
+
+  **Note:** Instead of statically defining these override blocks they can be passed in as a list of maps containing the above keys and their values. This allows dynamic, programmatic generation of overrides based on variable or environment data.
```
Again same 0.12 problem as above. But particularly for this one, I wrote all of this specifically for the feature of `dynamic, programmatic generation of overrides based on variable or environment data` which now feels defunct thanks to the new config syntax and particularly the foreach loops.

sandangel commented:
This PR takes > 1 year to review and becoming stale. I'd like to take this PR and do the rebase first ready for review.

arwalls commented:
What will it take to get this merged? We have the same need as many other commenters here.

brianpham commented:
Is anyone still working on this feature? Is there anyone else that can take over? I would love to be able to manage our spot fleet launch templates through Terraform.

japanfred commented:
I'm surprised this isn't getting more comments. This feature would be fantastic.

bflad commented:
Hi folks 👋 Please note that the commits in this pull request have been added into https://github.com/terraform-providers/terraform-provider-aws/pull/12732 and that PR is now under review. Thanks to @lossanarch for working on this implementation. Please follow that #12732 for further updates. 👍
